File: services/core-api/internal/httpapi/image_upload.go
package httpapi

import (
	"context"
	"errors"

	"github.com/huongnguyenduc/lumin-studio/services/core-api/internal/api"
	"github.com/huongnguyenduc/lumin-studio/services/core-api/internal/proofstore"
)

// errImageUploadNotConfigured mirrors the payment-proof case: the endpoint was hit but no lumin-assets
// image signer was wired (nil store). It maps to a generic 500 — the client cannot fix it and no signable
// contract exists — rather than leaking config state or issuing a partial upload form.
var errImageUploadNotConfigured = errors.New("httpapi: image uploads not configured")

// CreateImageUpload handles POST /uploads/image (P3-t t-6, P3-l). It returns a short-lived presigned POST
// form for one PERMANENT public image — a pet-page photo or a product-gallery photo — plus the host-pinned
// finalUrl the caller stores. Identical browser→Garage flow to CreatePaymentProofUpload (core-api only
// signs the policy, never proxies the body), but the object lands in the world-readable lumin-assets bucket
// with NO retention sweeper, so a lost pet's photo is never swept by payment-proof retention (t-6). Public +
// rate-limited: it mints nothing money-bearing, only an upload contract, so a session is not required but
// abuse is throttled.
func (s *Server) CreateImageUpload(ctx context.Context, req api.CreateImageUploadRequestObject) (api.CreateImageUploadResponseObject, error) {
	if req.Body == nil {
		return createImageUploadBadRequest(), nil
	}
	if s.imageUploads == nil {
		return nil, errImageUploadNotConfigured
	}
	if !s.imageUploadLimiter.allow() {
		if s.logger != nil {
			s.logger.Warn("image upload rate-limited")
		}
		return nil, errRateLimited
	}
	up, err := s.imageUploads.PresignPost(ctx, string(req.Body.ContentType))
	if errors.Is(err, proofstore.ErrInvalidContentType) {
		return createImageUploadBadRequest(), nil
	}
	if err != nil {
		return nil, err
	}
	return api.CreateImageUpload200JSONResponse(api.ImageUpload{
		UploadUrl: up.UploadURL,
		Fields:    up.Fields,
		FinalUrl:  up.FinalURL,
		ExpiresAt: up.ExpiresAt,
		MaxBytes:  up.MaxBytes,
	}), nil
}

func createImageUploadBadRequest() api.CreateImageUpload400JSONResponse {
	return api.CreateImageUpload400JSONResponse{BadRequestJSONResponse: api.BadRequestJSONResponse(envelope(codeValidation))}
}

File: apps/storefront/src/lib/image-upload.ts
'use server';

import { createApiClient } from '@lumin/api-client';
import type { components } from '@lumin/api-client';
import { coreApiBaseUrl } from './core-api';

// Server-side bridge for uploading ONE permanent public image — a pet-page photo (P3-t) — to the world-
// readable lumin-assets bucket via a presigned POST. The browser then POSTs the bytes STRAIGHT to Garage
// (core-api never proxies the file). Mirrors createPaymentProofUpload (lib/order-submit.ts) but hits
// POST /uploads/image, whose object has NO retention sweeper — so a lost pet's photo never expires (t-6),
// unlike a receipt on the 90-day proof bucket. Runs as a Server Action (CORE_API_URL is server-only) and
// persists nothing.

export type ImageUploadContentType = components['schemas']['ImageUploadInput']['contentType'];
export type ImageUpload = components['schemas']['ImageUpload'];

/** Ask core-api for a short-lived presigned POST form for ONE image of the given MIME type. On success the
 *  browser POSTs the file to `upload.uploadUrl` and keeps `upload.finalUrl`. Any failure collapses to a
 *  generic retryable `error`. */
export async function createImageUpload(
  contentType: ImageUploadContentType,
): Promise<{ ok: true; upload: ImageUpload } | { ok: false; code: 'error' }> {
  try {
    const client = createApiClient({ baseUrl: coreApiBaseUrl() });
    const { data } = await client.POST('/uploads/image', {
      body: { contentType },
    });
    if (data) return { ok: true, upload: data };
    return { ok: false, code: 'error' };
  } catch {
    return { ok: false, code: 'error' };
  }
}
